keps/sig-node/4381-dra-semantic-parameters/README.md

```diff
@@ -508,49 +445,35 @@ vendorParameters:
     uid: foobar-uid
   ...
   spec:
-    count: 2
     minimumRuntimeVersion: v12.0.0
-    minimumMemory: 8Gi
+    minimumMemory: 32Gi
     sharing:
       strategy: TimeSliced
 
 requests:
 - driverName: cards.dra.example.com
-  partitionable:
+  namedResourcesWithAttributes:
```

Comment:
Assume we have multiple models, rather than one which evolves.  E.g. we have `namedResourcesWithAttributes` and `partitionableResourcesWithHierarchicalAttributes`.  The driver+kubelet has to pick one of those to publish in `NodeResourceSlice`, right?

Then whoever produces the `ResourceClaimParameters` (which is still a request) needs to write the request in the same model (unless we guarantee the models are all transliterable), right?

Is this what the comment above was referring to, being"same version"?  

I don't think this flies, so I hope I misunderstand.  If I am doing a cluster update of the driver, I am going to have a point in time where some older drivers are publishing `namedResourcesWithAttributes` and new ones are publishing `partitionableResourcesWithHierarchicalAttributes`, and claims which are probably in `namedResourcesWithAttributes` form.  Won't that explode?

We already have a way of defining multiple schemata of transliterable data - API versions.  I don't think we want to depend on explicit sync of the models across actors, types, and time.  We don't usually micro-version internal structures within an API version (though I sometimes wish we could...).

But I keep coming back to "let's do less".  https://go.dev/play/p/gdnT0xef7iD  do we need this aspect (ResourceClaimParams) at all?  I'm not convinced.

Reply:
> Is this what the comment above was referring to, being"same version"?

Is that the comment about kubelet and DRA driver having to agree? No, that's not it. That refers to kubelet and DRA driver having to support the same gRPC interface which exchanges "resource.k8s.io/v1alpha2 namedResourcesWithAttributes" values. We can version that gRPC interface to support future versions.

> Won't that explode?

We would have to make this so that in resource.k8s.io/v1alpha3 we have partitionableResourcesWithHierarchicalAttributes and in resource.k8s.io/v1alpha2 we have namedResourcesWithAttributes, then convert between them. If we can define such a conversion, then we are fine. If we cannot, then we are defining a new, separate model. Those then co-exist, but drivers and app deployments have to change to the new one.

> We already have a way of defining multiple schemata of transliterable data - API versions. ... We don't usually micro-version internal structures within an API version (though I sometimes wish we could...).

My initial proposal had RawExtension to represent the model-specific data. The intend was exactly this "micro-versioning of the internal structures" because it would have decoupled the container API versioning (resource.k8s.io) from the versioning of the model. That was rejected because it would have meant persisting data without validation. I'm not sure whether you are suggesting to bring that back.

As it stands now, the semantic models get versioned through the resource.k8s.io API version.

Reply:
Versioning with API versions is OK.  Obviously, once we go beta it must be round-trippable, but until then we have some freedom break things.  What we can't do is have two different fields, each of which represents a distinct model which express incompatible things which the claim has to match exactly.

Reply:
> But I keep coming back to "let's do less". https://go.dev/play/p/gdnT0xef7iD do we need this aspect (ResourceClaimParams) at all? I'm not convinced.

Without claim parameters how do I do something like the following:
```
apiVersion: gpu.resource.nvidia.com/v1alpha1
kind: GpuClaimParameters
metadata:
  name: single-gpu
spec:
  count: 3
  sharing:
    strategy: MPS
    mpsConfig:
      maxConnections: <int>
      activeThreadPercentage: <int>
      pinnedDeviceMemoryLimit: <quantity>
```

Under the hood, the node will start a special process called an MPS server that governs the three GPUs that have been selected as part of this claim and sets the MPS server's properties to the `mpsConfig` provided. (thus controlling the way the GPUs are shared amongst the processes launched on them).

Reply:
> But I keep coming back to "let's do less". https://go.dev/play/p/gdnT0xef7iD do we need this aspect (ResourceClaimParams) at all? I'm not convinced.

So putting the selector into the ResourceClaim directly, together with something for quantities.
I have multiple problems with this very simplistic approach:
- cannot request different resources from different drivers
- cannot request resources with different selectors
- no support for configuration parameters

But most of all, there is no versioning of these requests. If the meaning or name of some attribute changes, then there's no way to figure out which revision of the semantic the user meant to use.

It also assumes that all models will have requests that can be described like this (CEL expression + named quantities). I think it's too early to make that decision. Can we first try with model-specific requests and then figure out whether some common pattern emerges?

Reply:
+1

If we can truly find a model in the future that isn't vendor specific and can cover 85% of the use cases, I'm all for it. I just don't think we have any idea what that looks like yet and we will never get there if we don't start somewhere.

Reply:
> But - we should set the expectation that alpha API users will be burned, at least for the modeling bits.

Agreed. This definitely is alpha and I wouldn't want to invest much effort into preserving compatibility either.

Reply:
OK, after discussing, I can buy how ClaimParameters is part of a complete story.

Revised: https://go.dev/play/p/4-FdpEfkcWg

> * cannot request different resources from different drivers
> * cannot request resources with different selectors

Is that something that belongs in a single claim?  If it were multiple drivers, do they operate independently or together?  

> But most of all, there is no versioning of these requests. If the meaning or name of some attribute changes, then there's no way to figure out which revision of the semantic the user meant to use.

I hear that concern, but also...don't do that.  If you change the semantic of a field with the same value, you kind of deserve to have angry users. :)

It also assumes that all models will have requests that can be described like this (CEL expression + named quantities). I think it's too early to make that decision. Can we first try with model-specific requests and then figure out whether some common pattern emerges?

> It also assumes that all models will have requests that can be described like this (CEL expression + named quantities).

It does, yes.  That said, The more we look at this, the more I disbelieve that any such things exists (or if so, they would probably be better if decomposed to fit this).

> I think it's too early to make that decision. Can we first try with model-specific requests and then figure out whether some common pattern emerges?

If we do that, just know that I will push HARD on this in the next stage, at which point it might become even more expensive to pivot (but that will be a you-problem, not a me-problem, and the kids say :).  

I'm worried that the longer we leave the bandage on, the more reluctant we are to pullit off, and the more residue will be left behind when we do.

Reply:
> > * cannot request different resources from different drivers
> > * cannot request resources with different selectors

> Is that something that belongs in a single claim? If it were multiple drivers, do they operate independently or together?

In the original DRA design, this was our "answer" to doing cross-device topology alignment. A meta driver could be built that consisted of *only* a controller, and defined a ClaimParameters object that could request two different resources that needed to be aligned. So long as that controller could speak the language of the drivers used to govern those resources at the kubelet level (what we are now standardizing on as the `NodeResourceSlice`) then it could run whatever custom logic it needed to make sure that it allocated those resources in an aligned way.

How this translates to the non-custom-controller world, I'm not sure yet.

Reply:
I don't think that's unsolvable in the reductive model.

Considering the case of a GPU with a NIC, we could:
* Write 2 claims (bad for lots of reasons)
* Make claims be a list (easy enough, uncertain side effects)
* Make the ClaimsParam include `wouldYouLikeASideOfNIC bool`
* Make the "with a NIC" be part of the modelled device (similar to DRA of today)
* Include enough topology in the model that a scheduler would know how to DTRT

Reply:
> > It also assumes that all models will have requests that can be described like this (CEL expression + named quantities).

> It does, yes. That said, The more we look at this, the more I disbelieve that any such things exists (or if so, they would probably be better if decomposed to fit this).

Even with what we've been proposing up until now (i.e. using vendor-specific CRDs and translating into CEL), I'm struggling with the knowledge that it is going to be difficult to do GPU-specific alignment (forget cross-device alignment) without some very complicated CEL expression.

Here is the algorithm that needs to run (which is dependent on the request size):
https://github.com/NVIDIA/go-gpuallocator/blob/main/gpuallocator/besteffort_policy.go#L20

My "plan" since we've been discussing this was to run this algorithm once at driver startup, build a set of lookup-tables encoded as attributes attached to each device, and then make sure I construct a CEL expression to do the lookup properly based on the incoming request size (and its other constraints).

I clearly don't want to leave such a thing up to the end-user, and it's so vendor specific that I can't imagine we would want to somehow embed this logic in k8s itself.

Reply:
Meaning something as seeminly simple as:
```
resourceClaimSpec:
	resourceClassName: "gpu.nvidia.com"
	count: 2
```
Is already already difficult to model if you want the allocated GPUs to be aligned on NVLINKs (which you *always* do, and I would try to enforce without even being requested by a user).

Reply:
Refining the options I scribbled:

1) The device modelling is rich enough to include information enough for the scheduler to recognize a pair with lowest cost (e.g. the old ACPI SLIT hack)
2) Expose attributes that let the user express the constraint, and allow compound claims.  E.g. I want 2 devices both have the same value for attribute "nvlink-id" or something.  CEL being extensible means we can encode common functionality as helper functions
3) Keep it opaque and let the node do last-mile assignment.  Sometimes that will fail.  Autoscaling can assume that any node which satisfies "has 2 GPUs" must have at least one pair which is co-usable.

I'm not sure *any* non-opaque model can do this easily, right?